src/relax/backend/vm/rewrite_cuda_graph.cc

```diff
@@ -0,0 +1,433 @@
+    ScopeInfo scope;
+    std::swap(scope, scope_);
+    for (const auto& binding : block->bindings) {
+      VisitBinding(binding);
+    }
+    // Summarize the detected static regions since we reach the end of the scope.
+    SummarizeStaticRegion();
+    std::swap(scope, scope_);
+    if (scope.static_regions.size()) {
+      block_to_static_regions_.emplace(block, std::move(scope.static_regions));
+    }
+  }
+
+  void VisitBinding_(const MatchCastNode* binding) final {
+    ExprVisitor::VisitBinding_(binding);
+    SummarizeStaticRegion();
+  }
+
+  bool IsStaticAlloc(const CallNode* alloc_tensor_call) {
+    auto shape = Downcast<ShapeExpr>(alloc_tensor_call->args[0]);
+    return std::all_of(shape->values.begin(), shape->values.end(),
+                       [](const PrimExpr& expr) { return expr.as<IntImmNode>() != nullptr; });
+  }
+
+  void VisitBinding_(const VarBindingNode* binding) final {
+    ExprVisitor::VisitBinding_(binding);
+
+    const Expr& value = binding->value;
+    static const auto& alloc_tensor_op = Op::Get("relax.builtin.alloc_tensor");
+    if (value->IsInstance<VarNode>()) {
+      // var -> var re-binding is not needed as we expect the binding is canonicalized
+      return;
+    }
+    if (value->IsInstance<ConstantNode>()) {
+      scope_.graph.AddBinding(binding, false);
+      return;
+    }
+    if (const auto* call = value.as<CallNode>()) {
+      if (call->op == alloc_tensor_op && IsStaticAlloc(call)) {
+        scope_.graph.AddBinding(binding, true);
+        return;
+      } else if (const auto* gv = call->op.as<GlobalVarNode>()) {
+        auto func = mod_->Lookup(GetRef<GlobalVar>(gv));
+        if (!func->IsInstance<tir::PrimFuncNode>()) {
+          return;
+        }
+
+        std::function<bool(const Expr&)> f_is_arg_static = [&](const Expr& arg) -> bool {
+          if (const auto* var = arg.as<VarNode>()) {
+            return scope_.graph.bindings.count(var);
+          } else if (arg->IsInstance<IntImmNode>() || arg->IsInstance<FloatImmNode>() ||
+                     arg->IsInstance<ConstantNode>()) {
+            return true;
+          } else if (arg->IsInstance<TupleNode>()) {
+            auto tuple = arg.as<TupleNode>();
+            return std::all_of(tuple->fields.begin(), tuple->fields.end(), f_is_arg_static);
+          }
+          return false;
+        };
+        bool is_all_args_static =
+            std::all_of(call->args.begin(), call->args.end(), f_is_arg_static);
+        if (is_all_args_static) {
+          scope_.graph.AddBinding(binding, false);
+          for (const Expr& arg : call->args) {
+            if (const auto* var = arg.as<VarNode>()) {
+              scope_.graph.AddEdge(var, binding->var.get());
+            }
+          }
+          return;
+        }
+      }
+    }
+    SummarizeStaticRegion();
+  }
+
+  void PruneOutputNodes(DependencyGraph* graph, const std::unordered_set<const VarNode*> outputs) {
```

Comment:
I think currently the nodes should be a tensor to make the logic works, but we use a var to represent them, and vars generally can be tuples